crates/ruff/src/jupyter/notebook.rs

```diff
@@ -0,0 +1,291 @@
+use std::fs::File;
+use std::io::{BufReader, BufWriter};
+use std::iter;
+use std::path::Path;
+
+use log::debug;
+use serde::Serialize;
+use serde_json::error::Category;
+
+use ruff_diagnostics::Diagnostic;
+use ruff_python_ast::types::Range;
+
+use crate::jupyter::{CellType, JupyterNotebook, SourceValue};
+use crate::rules::pycodestyle::rules::SyntaxError;
+use crate::IOError;
+
+pub const JUPYTER_NOTEBOOK_EXT: &str = "ipynb";
+
+/// Jupyter Notebook indexing table
+///
+/// When we lint a jupyter notebook, we have to translate the row/column based [Location]
+/// to jupyter notebook cell/row/column.
+#[derive(Debug, Eq, PartialEq)]
+pub struct JupyterIndex {
+    /// Enter a row (1-based), get back the cell (1-based)
+    pub row_to_cell: Vec<usize>,
```

Comment:
is there are a smart way to convert usize -> u32? because clippy complains

> casting `usize` to `u32` may truncate the value on targets with 64-bit wide pointers